contracts/mirror/src/contract.rs

```diff
@@ -0,0 +1,484 @@
+    bond.backup = backup.clone();
+    bond.return_type = return_type.clone();
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "update_bond"),
+        attr("id", id.to_string()),
+        attr("receiver", receiver),
+        attr("backup", backup.unwrap_or_default()),
+        attr("return_type", return_type.to_string()),
+    ];
+    Ok(response("update_bond_state", CONTRACT_NAME, attrs))
+}
+
+pub fn execute_bond(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    info: MessageInfo,
+    receiver: String,
+    r#ref: Option<String>,
+    backup: Option<String>,
+) -> ContractResult<Response<NeutronMsg>> {
+    let Config {
+        core_contract,
+        prefix,
+        ..
+    } = CONFIG.load(deps.storage)?;
+    backup
+        .as_ref()
+        .map(|addr| deps.api.addr_validate(addr))
+        .transpose()?;
+    bech32::decode(&receiver).map_err(|_| ContractError::WrongReceiverAddress)?;
+    ensure!(receiver.starts_with(&prefix), ContractError::InvalidPrefix);
+    let coin = cw_utils::one_coin(&info)?;
+    let counter = COUNTER.load(deps.storage)?;
+    let id = counter + 1;
+    COUNTER.save(deps.storage, &id)?;
+    let attrs = vec![
+        attr("action", "bond"),
+        attr("id", id.to_string()),
+        attr("receiver", receiver.to_string()),
+        attr("ref", r#ref.clone().unwrap_or_default()),
+        attr(
+            "backup",
+            backup.as_ref().map(|x| x.to_string()).unwrap_or_default(),
+        ),
+    ];
+    BONDS.save(
+        deps.storage,
+        id,
+        &BondItem {
+            id,
+            receiver,
+            backup,
+            received: None,
+            amount: coin.amount,
+            return_type: drop_staking_base::state::mirror::ReturnType::default(),
+            state: drop_staking_base::state::mirror::BondState::default(),
+        },
+    )?;
+    let msg = SubMsg::reply_on_success(
+        WasmMsg::Execute {
+            contract_addr: core_contract,
+            msg: to_json_binary(&drop_staking_base::msg::core::ExecuteMsg::Bond {
+                receiver: None,
+                r#ref,
+            })?,
+            funds: vec![coin],
+        },
+        id,
+    );
+    Ok(response("bond", CONTRACT_NAME, attrs).add_submessage(msg))
+}
+
+pub fn execute_complete(
+    deps: DepsMut<NeutronQuery>,
+    env: Env,
+    _info: MessageInfo,
+    items: Vec<u64>,
+) -> ContractResult<Response<NeutronMsg>> {
+    let mut msgs = vec![];
+    let mut attrs = vec![attr("action", "complete")];
+    let Config {
+        source_port,
+        source_channel,
+        ibc_timeout,
+        ..
+    } = CONFIG.load(deps.storage)?;
+    for id in items {
+        let mut bond = BONDS.load(deps.storage, id)?;
+        attrs.push(attr("id", id.to_string()));
+        attrs.push(attr("return_type", bond.state.to_string()));
+        attrs.push(attr("coin", bond.received.clone().unwrap().to_string()));
```

Comment:
I would rather write a comment why this `.unwrap()` is safe: because this field is guaranteed to be `Some(...)` and is only optional because we reuse it between calls to `execute()` and `reply()`